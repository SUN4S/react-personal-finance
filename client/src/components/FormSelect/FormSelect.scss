.selectDropdown {
  position: relative;
  width: 100%;
  border-radius: 3px;

  select {
    option {
      &:active {
        background-color: var(--primary);
      }
    }
  }
  &:after {
    content: "";
    position: absolute;
    top: 50%;
    right: 8px;
    width: 0;
    height: 0;
    border-top: 5px solid #aaa;
    border-right: 5px solid transparent;
    border-left: 5px solid transparent;
  }
}
