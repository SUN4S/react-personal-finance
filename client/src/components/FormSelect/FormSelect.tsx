import { FormSelectProps } from "../../models/inputs";

export const FormSelect: React.FC<FormSelectProps> = ({
  labelFor,
  label,
  inputTestId,
  name,
  options,
  required,
  register,
}) => {
  return (
    <label htmlFor={labelFor} data-testid="formLabel">
      {label}
      <div className="selectDropdown">
        <select
          {...register(name, { required: required })}
          defaultValue={"Essentials"}
          data-testid={inputTestId}
        >
          {options.map((item, index) => {
            return (
              <option value={item} key={index}>
                {item}
              </option>
            );
          })}
        </select>
      </div>
    </label>
  );
};
