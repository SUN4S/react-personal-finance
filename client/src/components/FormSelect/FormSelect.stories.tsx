import { Meta, Story } from "@storybook/react/types-6-0";

import { FormSelect } from "./FormSelect";
import { FormSelectProps } from "../../models/inputs";
import { useForm } from "react-hook-form";

interface TemplateProps extends FormSelectProps {
  theme: string;
}

export default {
  title: "Components/Form Select",
  component: FormSelect,
} as Meta;

const Template: Story<TemplateProps> = (args) => {
  const { register } = useForm();
  return (
    <html data-theme={args.theme}>
      <body
        style={{
          width: "400px",
          backgroundColor: "var(--sidebar-bg)",
          padding: "8px",
        }}
      >
        <FormSelect
          labelFor={args.labelFor}
          label={args.label}
          name={args.name}
          inputTestId={args.inputTestId}
          options={args.options}
          required={args.required}
          register={register}
        />
      </body>
    </html>
  );
};

export const Default_Light = Template.bind({});
Default_Light.args = {
  theme: "light",
  labelFor: "name",
  label: "Enter your name: ",
  options: ["Dog", "Cat", "Boat"],
  name: "name",
  inputTestId: "select",
  required: true,
};

export const Default_Dark = Template.bind({});
Default_Dark.args = {
  theme: "dark",
  labelFor: "name",
  label: "Enter your name: ",
  name: "name",
  options: ["Dog", "Cat", "Boat"],
  inputTestId: "select",
  required: true,
};
