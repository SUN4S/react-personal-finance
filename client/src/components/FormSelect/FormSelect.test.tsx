import { cleanup, fireEvent, render } from "@testing-library/react";

import { FormProvider } from "react-hook-form";
import { FormSelect } from "./FormSelect";
import renderer from "react-test-renderer";

describe("Form Select", () => {
  afterEach(cleanup);
  const mockOption = ["Essentials", "Wants", "Culture", "Unexpected"];
  const mockRegister = jest.fn();

  it("Renders", () => {
    const { getByTestId } = render(
      <FormSelect
        labelFor="category"
        label="Expense Category"
        name="category"
        inputTestId="category"
        options={["Essentials", "Wants", "Culture", "Unexpected"]}
        required
        register={mockRegister}
      />
    );
    expect(getByTestId("formLabel")).toBeTruthy();
  });

  it("Render Label", () => {
    const { getByTestId } = render(
      <FormSelect
        labelFor="category"
        label="Expense Category"
        name="category"
        inputTestId="category"
        options={["Essentials", "Wants", "Culture", "Unexpected"]}
        required
        register={mockRegister}
      />
    );
    expect(getByTestId("formLabel")).toHaveTextContent("Expense Category");
  });

  it("Renders Label For", () => {
    const { getByTestId } = render(
      <FormSelect
        labelFor="category"
        label="Expense Category"
        name="category"
        inputTestId="category"
        options={["Essentials", "Wants", "Culture", "Unexpected"]}
        required
        register={mockRegister}
      />
    );
    expect(getByTestId("formLabel")).toHaveAttribute("for", "category");
  });

  it("Renders Select", () => {
    const { getByTestId } = render(
      <FormSelect
        labelFor="category"
        label="Expense Category"
        name="category"
        inputTestId="category"
        options={["Essentials", "Wants", "Culture", "Unexpected"]}
        required
        register={mockRegister}
      />
    );
    expect(getByTestId("category")).toBeTruthy();
  });

  it("Renders Options", () => {
    const { getByTestId } = render(
      <FormSelect
        labelFor="category"
        label="Expense Category"
        name="category"
        inputTestId="category"
        options={["Essentials", "Wants", "Culture", "Unexpected"]}
        required
        register={mockRegister}
      />
    );
    const options = getByTestId("category").querySelectorAll("option");
    expect(options[0]).toHaveTextContent("Essentials");
    expect(options[1]).toHaveTextContent("Wants");
    expect(options[2]).toHaveTextContent("Culture");
    expect(options[3]).toHaveTextContent("Unexpected");
  });

  it("Options Selectable", () => {
    const { getByTestId } = render(
      <FormSelect
        labelFor="category"
        label="Expense Category"
        name="category"
        inputTestId="category"
        options={["Essentials", "Wants", "Culture", "Unexpected"]}
        required
        register={mockRegister}
      />
    );
    fireEvent.change(getByTestId("category"), {
      target: { value: "Culture", name: "Culture" },
    });
    expect(getByTestId("category")).toHaveValue("Culture");
  });

  it("Matches Snapshot", () => {
    const tree = renderer.create(
      <FormSelect
        labelFor="category"
        label="Expense Category"
        name="category"
        inputTestId="category"
        options={["Essentials", "Wants", "Culture", "Unexpected"]}
        required
        register={mockRegister}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
