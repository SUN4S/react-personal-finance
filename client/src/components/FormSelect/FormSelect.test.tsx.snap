// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Select Matches Snapshot 1`] = `
<label
  data-testid="formLabel"
  htmlFor="category"
>
  Expense Category
  <div
    className="selectDropdown"
  >
    <select
      data-testid="category"
      defaultValue="Essentials"
    >
      <option
        value="Essentials"
      >
        Essentials
      </option>
      <option
        value="Wants"
      >
        Wants
      </option>
      <option
        value="Culture"
      >
        Culture
      </option>
      <option
        value="Unexpected"
      >
        Unexpected
      </option>
    </select>
  </div>
</label>
`;
